docs/05-internals/design/principles/feature-toggles.md

```diff
@@ -0,0 +1,281 @@
+  an expiration date, developer contact information, etc.
+
+- A feature toggle should have a version number, so
+  you can get different versions of the same feature.
+  
+- It should be possible to configure a toggle in `application.conf`
+  so that its setting cannot be overridden per request.
+
+- The design of feature toggles should avoid ambiguity and
+  try to prevent situations where clients might be surprised by
+  unexpected functionality. It should be clear what will change
+  when a client requests a particular toggle setting. Therefore,
+  per-request settings should require the client to be explicit
+  about what is being requested.
+
+## Design
+
+### Configuration
+
+### Base Configuration
+
+The base configuration of feature toggles is in `application.conf`
+under `app.feature-toggles`. Example:
+
+```
+app {
+    feature-toggles {
+        new-foo {
+            description = "Replace the old foo routes with new ones."
+
+            available-versions = [ 1, 2 ]
+            default-version = 1
+            enabled-by-default = yes
+            override-allowed = yes
+
+            expiration-date = "2021-12-01T00:00:00Z"
+
+            developer-emails = [
+                "A developer <a.developer@example.org>"
+            ]
+        }
+
+        new-bar {
+            description = "Replace the old bar routes with new ones."
+
+            available-versions = [ 1, 2, 3 ]
+            default-version = 3
+            enabled-by-default = yes
+            override-allowed = yes
+
+            expiration-date = "2021-12-01T00:00:00Z"
+
+            developer-emails = [
+                "A developer <a.developer@example.org>"
+            ]
+        }
+
+        fast-baz {
+            description = "Replace the slower, more accurate baz route with a faster, less accurate one."
+
+            available-versions = [ 1 ]
+            default-version = 1
+            enabled-by-default = no
+            override-allowed = yes
+
+            developer-emails = [
+                "A developer <a.developer@example.org>"
+            ]
+        }
+    }
+}
+```
+
+All fields are required except `expiration-date`.
+
+Since it may not be possible to predict which toggles will need versions,
+all toggles must have at least one version. (If a toggle could be created
+without versions, and then get versions later, it would not be obvious
+what should happen if a client then requested the toggle without specifying
+a version number.) Version numbers must be an ascending sequence of
+consecutive integers starting from 1.
+
+If `expiration-date` is provided, it must be an [`xsd:dateTimeStamp`](http://www.datypic.com/sc/xsd11/t-xsd_dateTimeStamp.html). All feature toggles
+should have expiration dates except for long-lived ops toggles like `fast-baz` above.
+
+`KnoraSettingsFeatureFactoryConfig` reads this base configuration on startup. If
+a feature toggle has an expiration date in the past, the application will not start.
```

Comment:
oh, this is a hard one. I wouldn't punish ops for developers not removing a toggle in time ;-) Also what should then happen, if the expiration date is passed while the app is deployed into production. I hope the app will not stop working? I think a log output would be sufficient.

Reply:
I changed it to:

```
the application will log the error message "SHOW-STOPPER!!!" and the server will then self-destruct
```

Just kidding. 😃